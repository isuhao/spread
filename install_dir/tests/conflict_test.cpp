#include "common.cpp"

int answer = -1;

struct Answer : AskHandle
{
  bool askWait(AskPtr ask)
  {
    StringAsk *s = StringAsk::handle(ask);
    assert(s);

    cout << "QUESTION: " << s->message << endl;

    if(answer < 0 || answer >= s->options.size())
      {
        cout << "ABORT!\n";
        s->abortJob();
        return true;
      }

    cout << "ANSWER: " << s->options[answer] << endl;

    s->select(answer);
    return false;
  }
};

struct MyTest : DirInstaller
{
  MyTest(const std::string &pref = "")
    : DirInstaller(own, rules, cache, pref) {}

  HashDir add, del;
  Hash::DirMap upgrade;

  void test(const std::string &w)
  {
    cout << "TESTING " << w << ":\n";
    if(answer >= 0) cout << "  Preselected answer is: " << answer << endl;
    print(add, "ADD");
    print(del, "DEL");
    print(upgrade, "UPGRADE");
    doLog = false;
    resolveConflicts(add, del, upgrade);
    doLog = true;
    cout << "---\n";
    print(add, "ADD");
    print(del, "DEL");
    add.clear();
    del.clear();
    upgrade.clear();
    resetAll();
    answer = -1;
    cout << endl;
  }

  void a(const std::string &file, const Hash &h)
  { add.insert(HDValue(h,file)); }

  void d(const std::string &file, const Hash &h)
  { del.insert(HDValue(h,file)); }

  void u(const std::string &file, const Hash &h, const Hash &old)
  {
    a(file,h);
    upgrade[file] = old;
  }

  void c(const std::string &file, const Hash &h)
  {
    cache.add(file, h);
  }

  void doJob()
  {
    test("Nothing");

    a("file", Hash("FILE"));
    test("Overwriting non-existing file");

    a("file", Hash("FILE"));
    c("file", Hash("FILE"));
    test("Overwriting identical file");

    d("file", Hash("FILE"));
    test("Deleting non-existing file");

    d("file", Hash("FILE"));
    c("file", Hash("FILE"));
    test("Deleting expected file");

    u("file", Hash("FILE"), Hash("FILE"));
    c("file", Hash("FILE"));
    test("Upgrading identical file");

    u("file", Hash("NEW"), Hash("OLD"));
    c("file", Hash("OLD"));
    test("Upgrading expected file");

    u("file", Hash("NEW"), Hash("OLD"));
    test("Upgrading missing file");

    a("file1", Hash("HASH1"));
    a("file2", Hash("HASH1"));
    a("file3", Hash("HASH3"));
    u("file4", Hash("HASH4"), Hash("OLD4"));
    u("file5", Hash("HASH1"), Hash("HASH1"));
    c("file2", Hash("HASH1"));
    c("file3", Hash("HASH3"));
    c("file4", Hash("OLD4"));
    c("file6", Hash("ABC"));
    d("file6", Hash("ABC"));
    d("file7", Hash("DEF"));
    c("NOT_RELEVANT", Hash("BLAHBLAH"));
    test("Multiple targets");

    cout << "\nSTARTING USER-ASK TESTS:\n\n";

    a("file", Hash("FILE"));
    c("file", Hash("OLD"));
    answer = 2;
    test("Overwriting unexpected file - KEEP");

    a("file", Hash("FILE"));
    c("file", Hash("OLD"));
    answer = 1;
    test("Overwriting unexpected file - OVERWRITE");

    a("file", Hash("FILE"));
    c("file", Hash("OLD"));
    answer = 0;
    test("Overwriting unexpected file - OVERWRITE + BACKUP");

    u("file", Hash("FILE"), Hash("OLD"));
    c("file", Hash("WRONG"));
    answer = 2;
    test("Overwriting unexpected file - KEEP");

    u("file", Hash("FILE"), Hash("OLD"));
    c("file", Hash("WRONG"));
    answer = 1;
    test("Overwriting unexpected file - OVERWRITE");

    u("file", Hash("FILE"), Hash("OLD"));
    c("file", Hash("WRONG"));
    answer = 0;
    test("Overwriting unexpected file - OVERWRITE + BACKUP");

    d("file", Hash("FILE"));
    c("file", Hash("WRONG"));
    answer = 1;
    test("Deleting wrong file - KEEP");

    d("file", Hash("FILE"));
    c("file", Hash("WRONG"));
    answer = 0;
    test("Deleting wrong file - KILL!");

    assert(!checkStatus());
    setDone();
  }
};

int main()
{
  MyTest test("test");
  own.asker = new Answer;
  test.run();
  return 0;
}
