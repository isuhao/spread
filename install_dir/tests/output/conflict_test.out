TESTING Nothing:
ADD:
DEL:
UPGRADE:
---
ADD:
DEL:

TESTING Overwriting non-existing file:
ADD:
  FILE file
DEL:
UPGRADE:
---
ADD:
  FILE file
DEL:

TESTING Overwriting identical file:
ADD:
  FILE file
DEL:
UPGRADE:
---
ADD:
DEL:

TESTING Deleting non-existing file:
ADD:
DEL:
  FILE file
UPGRADE:
---
ADD:
DEL:

TESTING Deleting expected file:
ADD:
DEL:
  FILE file
UPGRADE:
---
ADD:
DEL:
  FILE file

TESTING Upgrading identical file:
ADD:
  FILE file
DEL:
UPGRADE:
  file FILE
---
ADD:
  FILE file
DEL:

TESTING Upgrading expected file:
ADD:
  NEW file
DEL:
UPGRADE:
  file OLD
---
ADD:
  NEW file
DEL:

TESTING Upgrading missing file:
ADD:
  NEW file
DEL:
UPGRADE:
  file OLD
---
ADD:
  NEW file
DEL:

TESTING Multiple targets:
ADD:
  HASH1 file1
  HASH1 file2
  HASH1 file5
  HASH3 file3
  HASH4 file4
DEL:
  ABC file6
  DEF file7
UPGRADE:
  file4 OLD4
  file5 HASH1
---
ADD:
  HASH1 file1
  HASH1 file5
  HASH4 file4
DEL:
  ABC file6


STARTING USER-ASK TESTS:

TESTING Overwriting unexpected file - KEEP:
  Preselected answer is: 2
ADD:
  FILE file
DEL:
UPGRADE:
QUESTION: File 'file' already exists. Overwrite it?
ANSWER: Keep file
---
ADD:
DEL:

TESTING Overwriting unexpected file - OVERWRITE:
  Preselected answer is: 1
ADD:
  FILE file
DEL:
UPGRADE:
QUESTION: File 'file' already exists. Overwrite it?
ANSWER: Overwrite without backup
---
ADD:
  FILE file
DEL:

TESTING Overwriting unexpected file - OVERWRITE + BACKUP:
  Preselected answer is: 0
ADD:
  FILE file
DEL:
UPGRADE:
QUESTION: File 'file' already exists. Overwrite it?
ANSWER: Overwrite with backup (recommended)
---
ADD:
  FILE file
  OLD file.___backup___
DEL:

TESTING Overwriting unexpected file - KEEP:
  Preselected answer is: 2
ADD:
  FILE file
DEL:
UPGRADE:
  file OLD
QUESTION: File 'file' contains modifications. Overwrite anyway?
ANSWER: Keep file
---
ADD:
DEL:

TESTING Overwriting unexpected file - OVERWRITE:
  Preselected answer is: 1
ADD:
  FILE file
DEL:
UPGRADE:
  file OLD
QUESTION: File 'file' contains modifications. Overwrite anyway?
ANSWER: Overwrite without backup
---
ADD:
  FILE file
DEL:

TESTING Overwriting unexpected file - OVERWRITE + BACKUP:
  Preselected answer is: 0
ADD:
  FILE file
DEL:
UPGRADE:
  file OLD
QUESTION: File 'file' contains modifications. Overwrite anyway?
ANSWER: Overwrite with backup (recommended)
---
ADD:
  FILE file
  WRONG file.___backup___
DEL:

TESTING Deleting wrong file - KEEP:
  Preselected answer is: 1
ADD:
DEL:
  FILE file
UPGRADE:
QUESTION: Deleted file 'file' has changes. Delete anyway?
ANSWER: Keep file
---
ADD:
DEL:

TESTING Deleting wrong file - KILL!:
  Preselected answer is: 0
ADD:
DEL:
  FILE file
UPGRADE:
QUESTION: Deleted file 'file' has changes. Delete anyway?
ANSWER: Delete file
---
ADD:
DEL:
  FILE file

